Router rewrite for PSR-7

Xerkus opened the pull request:
Work in progress

Goal of the rewrite is to switch component to the native support of PSR-7.

Console routing and http request routing are too different. Common interface proved to be limiting and of little practical value: routes were not shared between the two and consumers had to accept one and outright reject another in most cases.
Next implementation of zend-mvc will be PSR-7 based as well and zend-mvc-console will be dropped.

For that reason, support for `Zend\Http\RequestInterface` will be dropped as too incompatible with PSR-7 and support for the non-http routing with `Zend\Stdlib\RequestInterface` will not be retained as impractical.
Effectively that will make zend-router a PSR-7 only router.

Also, this PR bumps minimum php to 7.1, introduces strict types, scalar and return typehinting

Note: this PR is a subject for rebases and commit squashes. Ask before using as a basis for your work.

Xerkus committed:
Bump minimum php to 7.1

Xerkus committed:
Fix or suppress type errors

Xerkus commented:
Small note:
I wanted to keep RouteMatch as a separate thing from RouteResult to keep it more or less backward compatible.
It is used extensively in zend-mvc, but consumers should really be using PSR-7 request attributes for matched parameters. While it will cause additional migration challenges for zend-mvc users, it will be more in line with zend-expressive and the rest of the industry.
Besides, attempts to keep RouteMatch and request attributes in sync are error prone and pretty much guaranteed to produce subtle bugs.

Xerkus commented:
Routes doing partial match conditionally and depending on the concrete implementation are violating `RouteInterface` contract.
To make things more straightforward, `RouteInterface` is now not allowed to return partial match. `PartialRouteInterface` extending `RouteInterface` was introduced instead. 
`Chain`, `Hostname`, `Literal`, `Method`, `Regex`, `Scheme` and `Segment` are now partial routes.
`Chain` was changed to accept partial routes only, `Part` was changed to accept only partial route for its base route.

Xerkus committed:
Remove unneeded composer conflict

Xerkus committed:
Be strict about unintentional coverage

webimpress commented on composer.json:
```diff
@@ -18,20 +18,18 @@
     "require": {
-        "php": "^5.6 || ^7.0",
+        "php": "^7.1",
         "container-interop/container-interop": "^1.2",
```
Maybe we can use psr/container (PSR-11), or we need first SMv4?

webimpress commented on src/Route/Regex.php:
```diff
@@ -54,7 +61,7 @@ class Regex implements RouteInterface
      * @param  array  $defaults
      */
-    public function __construct($regex, $spec, array $defaults = [])
+    public function __construct(string $regex, $spec, array $defaults = [])
```
Should we also have typehint on `$spec`? In PHPDocs I can see "string" only.

webimpress commented on src/Route/Scheme.php:
```diff
@@ -51,7 +58,7 @@ public function __construct($scheme, array $defaults = [])
      * @throws Exception\InvalidArgumentException
      */
-    public static function factory($options = [])
+    public static function factory($options = []) : RouteInterface
```
Can we have "iterable" type hint here?

webimpress commented on src/Route/Segment.php:
```diff
@@ -121,7 +128,7 @@ public function __construct($route, array $constraints = [], array $defaults = [
      * @throws Exception\InvalidArgumentException
      */
-    public static function factory($options = [])
+    public static function factory($options = []) : RouteInterface
```
"iterable" typehint here?

webimpress commented on src/Route/Segment.php:
```diff
@@ -154,7 +161,7 @@ public static function factory($options = [])
      * @throws Exception\RuntimeException
      */
-    protected function parseRouteDefinition($def)
+    protected function parseRouteDefinition($def) : array
```
"string" typehint on param

webimpress commented on src/Route/Segment.php:
```diff
@@ -229,7 +236,7 @@ protected function parseRouteDefinition($def)
      * @return string
      */
-    protected function buildRegex(array $parts, array $constraints, &$groupIndex = 1)
+    protected function buildRegex(array $parts, array $constraints, &$groupIndex = 1) : string
```
can we add type hint on the last param here, please?

webimpress commented on src/SimpleRouteStack.php:
```diff
@@ -64,7 +66,7 @@ public function __construct(RoutePluginManager $routePluginManager = null)
      * @throws Exception\InvalidArgumentException
      */
-    public static function factory($options = [])
+    public static function factory($options = []) : RouteInterface
```
Hm... why not "SimpleRouteStack" as return type in signature?

Xerkus replied:
not possible

Xerkus committed:
Add strict types declaration

Xerkus committed:
Bumped to next dev version (4.0.0)

Xerkus committed:
Add CHANGELOG for #44

Xerkus committed:
Drop zend-http and add diactoros

Diactoros should be replaced by PSR-17 implementation once it is
accepted

Xerkus committed:
Drop deprecated Wildcard route

Xerkus committed:
Move TreeRouteStack one namespace up

Xerkus committed:
Change RouteInterface to PSR-7

- Change RouteInterface::match() signature, it now requires PSR-7
  ServerRequestInterface and always returns RouteResult
- Change RouteInterface::assemble() signature, it now requires PSR-7
  UriInterface as first parameter and returns UriInterface. Usage of
  UriInterface allows for more flexibility in assembling uri from tree
  route parts as compared to simple string manipulation.
- Drop factory() from RouteInterface. Routes can be greatly simplified
  by moving the creation logic out.
- Introduce PartialRouteInterface extending RouteInterface to allow
  partial matching enforced by contract.
- Drop Http specific RouteInterface as zend-router is now Http only
- RouteStackInterface is extended with additional getters previously
  present only in SimpleRouteStack and by extension in TreeRouteStack.
  Fluent interface is removed.

Xerkus committed:
Introduce RouteResult, deprecate RouteMatch

RouteResult is an immutable value object representing routing result,
succesfull or otherwise. RouteResult replaces RouteMatch, which is now
deprecated.

Routing failure due to HTTP method should produce RouteResult with
method failure status and a list of allowed methods, which then can be
used to produce 405 response.

Xerkus committed:
Route stack refactor stub

Xerkus committed:
Drop translator aware route stack

Xerkus committed:
Rename Zend\Router\Http namespace to Zend\Router\Route: move files

Xerkus committed:
Rename Zend\Router\Http namespace to Zend\Router\Route

Xerkus committed:
Add PartialRouteTrait for common match() implementation

Xerkus committed:
Add helper class and traits for route testing

- RoutTestDefinition is a value object to programatically set used
  parameters and expected results for use by data providers defined in
  traits.
- RouteTestTrait provides route matching and uri assembling tests along
  with data providers for them
- PartialRouteTestTrait provides partial route matching test with data
  provider for it

Xerkus committed:
Refactor Literal route for PSR-7

Xerkus committed:
Refactor Hostname route for PSR-7

Xerkus committed:
Delete child route match

Xerkus committed:
Refactor Method route for PSR-7

Xerkus committed:
Refactor Scheme route for PSR-7

Xerkus committed:
Refactor Segment route for PSR-7

Xerkus committed:
Refactor Regex route for PSR-7

Xerkus committed:
Update PartialRouteResult and add allowed methods for successful match

Xerkus committed:
Make Method route provide allowed methods on match

Xerkus committed:
Refactor Part route for PSR-7

Xerkus committed:
Refactor Chain route for PSR-7

Xerkus committed:
Remove now redundant HttpRouterFactory

Xerkus committed:
Add route definitions to RoutePluginManager and update factory

Xerkus committed:
Initial implementation for route config factory

Xerkus committed:
Move chain workaround to chain route factory

Xerkus committed:
Add psr container to direct dependencies

Xerkus committed:
Remove prototypes from factory and accept as extra parameter instead

Xerkus committed:
Add Router class and its factory

Xerkus committed:
Reintroduce TranslatorAwareRouteStackDecorator inplace of dropped tree route stack

Xerkus committed:
Drop unused RouterFactory

Xerkus committed:
Improve tests for route plugin manager factory

Xerkus committed:
Increment assertion count for test not doing assertions by intention

Xerkus committed:
Remove service manager v2 support in RouteInvokableFactory

Xerkus committed:
Remove unused test assets

Xerkus committed:
Comment out and keep test for ambiguous use case having less desired behavior

Xerkus committed:
Remove obsolete RouterConfigTrait

Xerkus committed:
Apply cs fixes

Xerkus committed:
Update copyright headers

Xerkus committed:
Add RouteConfigFactory factory

weierophinney commented:
This repository has been closed and moved to laminas/laminas-router; a new issue has been opened at https://github.com/laminas/laminas-router/issues/6.

weierophinney commented:
This repository has been moved to laminas/laminas-router. If you feel that this patch is still relevant, please re-open against that repository, and reference this issue. To re-open, we suggest the following workflow:
 - Squash all commits in your branch (`git rebase -i origin/{branch}`)
 - Make a note of all changed files (`git diff --name-only origin/{branch}...HEAD
 - Run the laminas/laminas-migration tool on the code.
 - Clone laminas/laminas-router to another directory.
 - Copy the files from the second bullet point to the clone of laminas/laminas-router.
 - In your clone of laminas/laminas-router, commit the files, push to your fork, and open the new PR.
 We will be providing tooling via laminas/laminas-migration soon to help  automate the process.
